### PR TITLE
feat: Add npm builder workflow

### PR DESCRIPTION
**Pull request by laurentsimon**
This PR adds a new workflow `builder_node_slsa3.yml`. It contains the structure to build upon for follow-up PRs that will create the node builder itself.

### CODE REVIEW

File: .github/workflows/builder_node_slsa3.yml [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+    inputs:
+      node-version:
+        description: >
+          The node version to use.
+
+          This argument is passed, unchanged, to `actions/node-setup`'s
+          as `node-version` parameter.
+        required: true
+        type: string
+      # TODO: support uploading assets to GitHub release.
```
Comment (ianlewis) [inline .github/workflows/builder_node_slsa3.yml:55]: nit: Consider adding a github issue and adding the URL to the comment. Same for other TODOs.

Comment (ianlewis) [reply .github/workflows/builder_node_slsa3.yml:55]: We should also add a new milestone and add these issues to it.

File: .github/workflows/builder_node_slsa3.yml [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+          # npm ci <args>
+          ./"$BUILDER_BINARY" ci \
+            --ci-arguments "$UNTRUSTED_CI_ARGUMENTS" \
+            --directory "$UNTRUSTED_DIR"
+
+      # TODO(hermeticity) OS-level.
+      # - name: Disable hermeticity
+      # uses: slsa/hermeticity@xxx
+      #   with:
+      #     to-state: enabled
```
Comment (ianlewis) [inline .github/workflows/builder_node_slsa3.yml:240]: nit: Maybe we can omit this comment?

File: .github/workflows/builder_node_slsa3.yml [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+# limitations under the License.
+
+name: SLSA node builder
+
+permissions:
+  contents: read
+
+env:
+  # Builder.
+  BUILDER_BINARY: slsa-builder-go-linux-amd64 # Name of the binary in the release assets.
```
Comment (ianlewis) [inline .github/workflows/builder_node_slsa3.yml:22]: do you really want to use the Go builder when pulling a release?

File: .github/workflows/builder_node_slsa3.yml [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+          path: "${{ steps.sign-prov.outputs.signed-provenance-name }}"
+          if-no-files-found: error
+          retention-days: 5
+
+  ###################################################################
+  #                                                                 #
+  #           Publish the package and its provenance                #
+  #                                                                 #
+  ###################################################################
+  publish:
```
Comment (ianlewis) [inline .github/workflows/builder_node_slsa3.yml:344]: I'm guessing we should probably have an option that allows you to skip the publishing step.

File: .github/workflows/builder_node_slsa3.yml (hunk 1 above)
Comment (ianlewis) [reply .github/workflows/builder_node_slsa3.yml:55]: I added a [milestone](https://github.com/slsa-framework/slsa-github-generator/milestone/8) that we can add future issues to and can track publicly as part of the roadmap.

File: .github/workflows/builder_node_slsa3.yml [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+
+      - name: Download dependencies
+        env: 
+          UNTRUSTED_CI_ARGUMENTS: "${{ inputs.ci-arguments }}"
+          UNTRUSTED_DIR: ${{ inputs.directory }}"
+        run: |
+          set -euo pipefail
+
+          # npm ci <args>
+          ./"$BUILDER_BINARY" ci \
```
Comment (ianlewis) [inline .github/workflows/builder_node_slsa3.yml:232]: I wonder if we can't combine the `ci` and `build` steps. especially if we will always be running `npm ci`

File: .github/workflows/builder_node_slsa3.yml (hunk 1 above)
Comment (laurentsimon) [reply .github/workflows/builder_node_slsa3.yml:55]: yeah I was planning on creating an issue for it. Thanks for the reminder.

File: .github/workflows/builder_node_slsa3.yml (hunk 3 above)
Comment (laurentsimon) [reply .github/workflows/builder_node_slsa3.yml:22]: oversight. Fixed.

File: .github/workflows/builder_node_slsa3.yml (hunk 4 above)
Comment (laurentsimon) [reply .github/workflows/builder_node_slsa3.yml:344]: How about we don't provide it right away and see if people ask for it? By default, this option would be set to `true`, so we could add it later. Wdut?

File: .github/workflows/builder_node_slsa3.yml (hunk 5 above)
Comment (laurentsimon) [reply .github/workflows/builder_node_slsa3.yml:232]: One reason (maybe not a strong one) is if we want to turn on some OS-level hermeticity after downloading the dependencies in the future.

File: .github/workflows/builder_node_slsa3.yml (hunk 2 above)
Comment (laurentsimon) [reply .github/workflows/builder_node_slsa3.yml:240]: I shortened it a to a single line. Unless we think it would be done in the builder. Let me know.

**Comment by laurentsimon** [comment]
friendly ping

**Review by ianlewis — APPROVED** [review]

**Pull request merged**